feat: add autocomplete for project name arguments, fixes #4880

GuySartorelli opened the pull request:
<!-- 
  PR titles have very precise rules, please read 
  https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-pull-requests
-->

## The Issue

Lots of the go commands take a project name as an argument, but don't provide autocompletion for those arguments.

## How This PR Solves The Issue

Adds autocompletion for commands that take a project name as an argument.

- If only one argument is allowed, it doesn't give completion for more.
- If multiple arguments are allowed, it gives completion for all of them.
- For `start` and `stop` specifically, if only gives names of the relevant (inactive and active respectively) projects - for all other commands, all names are given.

**NOTE** This makes the `--select` flag for `ddev start` and `ddev stop` redundant. I've marked the flag as hidden and deprecated for both of those commands - it can be removed in some future release.

I've also removed the docs about that flag - but if you want me to add it back with a note that they're deprecated, I can happily do that.

## Manual Testing Instructions

Type `ddev start <tab>` and see the autocompletion results.
- Try this with any command that takes a project name as an argument
- Check that commands that only accept one argument only perform completion for one argument
- Check that commands that accept multiple arguments perform completion for as many arguments as you like
- Check that `ddev start` only gives completion for currently inactive projects
- Check that `ddev stop` only gives completion for currently active projects
- Check that a project name isn't repeated if it's already included as an argument

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->
I've added tests that work based on the `TestSites` array - though since these tests won't run with less than 2 test sites, the tests are skipped if there's not enough tests.

## Related Issue Link(s)

- https://github.com/ddev/ddev/issues/4880

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->
This is fully backwards compatible and shouldn't affect anything else.

GuySartorelli commented:
Whitespace changes and imports moving around were done automatically by my IDE

stasadev commented on pkg/ddevapp/utils.go:
```diff
@@ -400,6 +401,44 @@ func ExtractProjectNames(apps []*DdevApp) []string {
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+		return ExtractProjectNames(apps), cobra.ShellCompDirectiveNoFileComp
```
```suggestion
		// Do not show arguments that are already written on the command line
		var appNames []string
		appNamesFromCmd := make(map[string]struct{})
		for _, name := range cmd.Flags().Args() {
			appNamesFromCmd[name] = struct{}{}
		}
		for _, name := range ExtractProjectNames(apps) {
			if _, exists := appNamesFromCmd[name]; !exists {
				appNames = append(appNames, name)
			}
		}
		return appNames, cobra.ShellCompDirectiveNoFileComp
```

stasadev replied:
The suggestion is a fix for:

- `ddev start project-name <TAB>` and `project-name` is still in the list.
- one project is running and you run `ddev stop <TAB><TAB><TAB>`, the result is `ddev stop project-name project-name project-name`.
- and so on.

GuySartorelli replied:
Ooh good catch.

GuySartorelli replied:
I did something a little different than the suggestion, but the result is the same.
- Called the variable `projectNames` instead of `appNames` since that's the terminology used in the rest of the function
- Used the `args` argument that cobra is giving us instead of fetching the args manually.

stasadev replied:
Thanks, that looks better 👍, because I just googled how to do `array_intersect()` in go and used the first option I found.

stasadev reviewed (changes requested):
Wow, this is HUGE improvement 🎉

> This makes the `--select` flag for `ddev start` and `ddev stop` redundant

We can mark that flag as deprecated and hidden, and remove it later.

> should I create a new `command_completion_test.go` and do a few tests in there?

I think a new file with tests is suitable for this.

Thank you, @GuySartorelli 👍

GuySartorelli commented:
I've marked the `--select` flag for `ddev start` and `ddev stop` as hidden and deprecated.

I've also made a start on tests (which will currently fail) - I need to see if there's a tidy way to either:
- a) only create a few specific sites so I can rely on specific site names
- b) write the tests in a way that generically uses names of whatever test sites happen to be spun up

I ran out of time to do that tonight. Any pointers would be welcome.
Marking as draft for now until I get them sorted.

stasadev commented:
> a) only create a few specific sites so I can rely on specific site names

My opinion: the only requirement is that the number of sites should be sufficient to test all possible cases.

GuySartorelli commented:
I've added tests now. They work using the existing `TestSites` array because I ran into weird problems when I tried to define my own. The only caveat there is that the tests will be skipped if there's less than 2 tests in the array (which if I'm not mistaken will only happen if someone is testing locally and uses the `GOTEST_SHORT` env variable).

I also removed the docs about the deprecated/hidden `--select` flag but if you want to keep that doc around I can bring it back with a note about it being deprecated. Makes more sense to me to just not mention it in the docs though.

I've also updated the PR description.

stasadev reviewed (approved):
Wow, thank you so much for this!

LGTM, tested and checked the list of changed commands.

rfay commented:
You are completely amazing @GuySartorelli . I won't have time to adequately grok this for at least a few days, but so appreciative! This will be a much-loved feature.

rfay commented:
Rebased, reviewing.

@GuySartorelli please see if you can convince your IDE to use standard gofmt formatting, which is what's generally required. As you point out, it makes a mess and is hard to review when things get changed that aren't relevant to the PR. Happy to chat with you about this, I know the timezones are hard. But it would be cool if you could experiment a bit to see if you can calm it down a bit. What IDE are you using? How do you have it set up? We can discuss in another issue if that's the easiest.

Thanks so much for your incredible contributions. You know we're looking forward to more. I'm going to learn a lot from you.

rfay commented on cmd/ddev/cmd/debug-compose-config.go:
```diff
@@ -12,8 +13,9 @@ import (
 var DebugComposeConfigCmd = &cobra.Command{
-	Use:   "compose-config [project]",
-	Short: "Prints the docker-compose configuration of the current project",
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("all", 1),
```
I imagine you already explored to see if this could be handled somewhere other than each Cobra Cmd.

rfay reviewed (changes requested):
This is fantastic and creative work.

Right now, I can't get it to work at all. When I click `ddev start <tab>` I just get a list of filenames in current directory. This is with bash 3, macOS.  What do I need to do to manually test it?

rfay commented on cmd/ddev/cmd/logs.go:
```diff
@@ -13,9 +14,10 @@ var (
-	Use:   "logs [projectname]",
-	Short: "Get the logs from your running services.",
-	Long:  `Uses 'docker logs' to display stdout from the running services.`,
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("all", 1),
```
Should only allow running projects I think.

GuySartorelli replied:
Fixed

rfay reviewed (changes requested):
This is looking great and ready to go. I did find a couple of things that should only show running projects, please take a casual look and see if there are others. thanks for the amazing work!

GuySartorelli commented:
> please see if you can convince your IDE to use standard gofmt formatting, which is what's generally required.

If you can point me to something that explains what that specifically means, I'll be happy to

GuySartorelli commented on cmd/ddev/cmd/clean.go:
```diff
@@ -14,8 +14,9 @@ import (
 var CleanCmd = &cobra.Command{
-	Use:   "clean [projectname ...]",
-	Short: "Removes items DDEV has created",
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("all", 0),
```
Oooh I'll look further into that (separately), that seems useful.

rfay commented:
Github discussion is great. Some vscode plugins use rather exotic and opinionated golang setup, whereas we're trying mostly to just use what `gofmt` says to do, so `gofmt -w -s` is what's expected everywhere. Your IDE should be able to do that. That said, I didn't really see all that much to worry about in this PR.

GuySartorelli commented on cmd/ddev/cmd/ssh.go:
```diff
@@ -15,9 +15,10 @@ var sshDirArg string
-	Use:   "ssh [projectname]",
-	Short: "Starts a shell session in the container for a service. Uses web service by default.",
-	Long:  `Starts a shell session in the container for a service. Uses web service by default. To start a shell session for another service, run "ddev ssh --service <service>`,
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("all", 1),
```
Fixed

rfay commented:
This is working wonderfully for me in testing just now.
